### PR TITLE
[WIP] WMS Time slider component

### LINKED ISSUE #1241: WMS-time support
Spec: https://github.com/camptocamp/c2cgeoportal/wiki/Spec-%231708-Layertree-desktop#time-data
Doc: https://camptocamp.github.io/c2cgeoportal/master/integrator/wmstime.html

 * single date with a slider
 * single date with a date picker
 * date range with a slider
 * date range with two date picker

When a slider is used either:
 * min and max with an interval (step) date must be supported
 * list of values (dates) must be supported

When a date picker is used, min and max date must be supported.

Standard html5 input range can't be used but jQuery UI seems to have everything we need:
 * http://jqueryui.com/datepicker/ (https://github.com/angular-ui/ui-date)
 * http://jqueryui.com/slider/ (https://github.com/angular-ui/ui-slider)

The "time" configuration comes from the theme service, example:
```
time: {
  widget: "datepicker",
  maxValue: "2013-12-31T00:00:00Z",
  minValue: "2006-01-01T00:00:00Z",
  maxDefValue: null,
  minDefValue: null,
  resolution: "day",
  mode: "range"
}
```

### PR DESCRIPTION
**Pull request by olivierSemet**
Close #1241 

Demo of the component: https://oliviersemet.github.io/ngeo/time-slider-component/examples/contribs/gmf/timeslider.html

Demo app: Currently the node with a time parameter set to 'slider' is not reachable but as soon as the demo server is updated with @sbrunner modifications, it should be available.

ping @fredj @sbrunner @pgiraud

### CODE REVIEW

**Comment by olivierSemet** [comment]
For the example no locale is provided so it uses the 'US' date format which does not shift month and day with 0 when it's needed. That's why the size of tooltip change a lot.
Two solutions:

1. Provide a local to the example
2. Fix the width of the tooltip

File: contribs/gmf/src/directives/timeslider.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        var el = $(this);
+        if (!el.attr('aria-describedby')) {
+          var dateText  = Array.isArray(ctrl.dates) ?
+          $filter('date')(new Date(ctrl.dates[i]), dateFormat) :
+          $filter('date')(new Date(ctrl.dates), dateFormat);
+
+          el.tooltip({
+            'selector': 'span.ui-slider-handle',
+            'trigger': 'manual',
+            'placement' : i === 0 && 'bottom' || 'top',
```
Comment (pgiraud) [inline contribs/gmf/src/directives/timeslider.js]: `i === 0 ? 'bottom' : 'top'` is easier to read in my opinion

Comment (pgiraud) [inline contribs/gmf/src/directives/timeslider.js]: Also, though I think I understand why you put one at the bottom and one at the top. This makes things really weird at first. I'd prefer showing the tooltips only on hover.

File: contribs/gmf/examples/timeslider.html [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      .tooltip {
+        min-width: 6.5rem;
+      }
+    </style>
+  </head>
+  <body ng-controller="MainController as ctrl">
+    <p id="desc">This example shows how to use the <code>gmf-time-slider</code> directive.</p>
+    <ul>
+      <li>
+        Select a periode with a timeslider
```
Comment (pgiraud) [inline contribs/gmf/examples/timeslider.html]: period

File: contribs/gmf/examples/timeslider.html [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        Date selected:
+        <pre>
+          {{ctrl.sliderRangeValue | json}}
+        </pre>
+      </li>
+      <li style="margin-top: 50px;">
+        Select a single date with a timeslider
+      </li>
+      <li>
+          <gmf-time-slider time="ctrl.slider" on-date-selected="ctrl.onDateSelected(time)"></gmf-time-slider>
```
Comment (pgiraud) [inline contribs/gmf/examples/timeslider.html]: 2 spaces identation

File: contribs/gmf/src/directives/timeslider.js (hunk 1 above)
Comment (olivierSemet) [inline contribs/gmf/src/directives/timeslider.js]: To handle that case:
capture d ecran 2016-06-09 a 14 38 38

File: contribs/gmf/less/mobilelayertree.less [hunk 4]
```diff
@@ -21,4 +21,8 @@ nav.nav-left .gmf-layertree-node a[data-toggle] {
   .group {
     font-size: @font-size-base;
   }
+
+  [ngeo-popover] {
+    display: none;
+  }
```
Comment (pgiraud) [inline contribs/gmf/less/mobilelayertree.less]: Why is required? Does it have any relationship with the current pull request?

File: contribs/gmf/src/directives/timeslider.js [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+     */
+    function(element, attrs) {
+      var templateUrl = attrs['gmfTimeSliderTemplateUrl'];
+      return templateUrl !== undefined ? templateUrl :
+          gmf.baseTemplateUrl + '/timeslider.html';
+    });
+
+
+/**
+ * Provide a directive to select a signle date or a range of dates with a slider
```
Comment (pgiraud) [inline contribs/gmf/src/directives/timeslider.js]: single

File: contribs/gmf/src/directives/timeslider.js [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    templateUrl : gmfTimeSliderTemplateUrl,
+    link: function(scope, element, attrs, ctrl) {
+
+      /** @type {string} */
+      var dateFormat = 'shortDate';
+
+      ctrl.sliderOptions['slide'] = updateTooltip_;
+      ctrl.sliderOptions['start'] = updateTooltip_;
+      ctrl.sliderOptions['stop'] = onSliderReleased_;
+      $timeout(init_);
```
Comment (pgiraud) [inline contribs/gmf/src/directives/timeslider.js]: Can you please tell why you need a timeout here?

File: contribs/gmf/less/mobilelayertree.less (hunk 4 above)
Comment (olivierSemet) [inline contribs/gmf/less/mobilelayertree.less]: Not really I just found the problem while I was testing, the popover should not be display on mobile right?

File: contribs/gmf/src/directives/timeslider.js [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+            }
+          });
+        }
+
+        element.find('.ui-slider-handle').on('shown.bs.tooltip', function() {
+          //Display sibling (only for range time).
+          var otherHandle = $(this).siblings('.ui-slider-handle');
+          if (otherHandle.length > 0) {
+            //test if other handle has its toolip hidden (do not call show for nothing)
+            if (!$(otherHandle[0]).attr('aria-describedby')) {
```
Comment (pgiraud) [inline contribs/gmf/src/directives/timeslider.js]: What would be the cases where second handle is hidden?

File: contribs/gmf/src/directives/timeslider.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+              otherHandle.tooltip('show');
+            }
+          }
+        });
+
+        element.find('.ui-slider-handle').on('hide.bs.tooltip', function(jqEvt) {
+          if (element.find(this).length > 0) {
+            jqEvt.preventDefault();
+          }
```
Comment (pgiraud) [inline contribs/gmf/src/directives/timeslider.js]: What is this for?

File: contribs/gmf/src/directives/timeslider.js [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        tooltip.css('left', left - tooltip.width() / 2);
+        tooltip.find('.tooltip-inner').text(dateText);
+      }
+
+      function initToolTip_(i) {
+        var el = $(this);
+        if (!el.attr('aria-describedby')) {
+          var dateText  = Array.isArray(ctrl.dates) ?
+          $filter('date')(new Date(ctrl.dates[i]), dateFormat) :
+          $filter('date')(new Date(ctrl.dates), dateFormat);
```
Comment (pgiraud) [inline contribs/gmf/src/directives/timeslider.js]: I'd prefer this:
```javascript
var date = Array.isArray(ctrl.dates) ? ctrl.dates[i] : ctrl.dates;
// or even
// var date = ctrl.dates[i] || ctrl.dates;
var dateText = $filter('date')(new Date(date), dateFormat);
```

File: contribs/gmf/src/directives/timeslider.js [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    'min' : min,
+    'max' : max,
+    'step' : step
+  };
+
+  /**
+   * @type {Array|number}
+   * @export
+   */
+  this.dates = this.time.mode === 'range' && [min, max] || Math.ceil((max - min) / 2);
```
Comment (olivierSemet) [inline contribs/gmf/src/directives/timeslider.js]: missing min:
Math.ceil((max - min) / 2)
should be Math.ceil(min + (max - min) / 2)

File: contribs/gmf/src/directives/timeslider.js (hunk 8 above)
Comment (olivierSemet) [inline contribs/gmf/src/directives/timeslider.js]: Prevent 'mouseleave/mouseout' event to hide the tooltip.

File: contribs/gmf/less/mobilelayertree.less (hunk 4 above)
Comment (pgiraud) [inline contribs/gmf/less/mobilelayertree.less]: Sorry to be picky but if this it's not related to current pull request, then it should be addressed in a different pull request or at least in a separate commit.

File: contribs/gmf/src/directives/timeslider.js (hunk 7 above)
Comment (olivierSemet) [inline contribs/gmf/src/directives/timeslider.js]: Yes this case can't happen anymore

**Comment by olivierSemet** [comment]
Thanks for review @pgiraud

File: contribs/gmf/less/mobilelayertree.less (hunk 4 above)
Comment (olivierSemet) [inline contribs/gmf/less/mobilelayertree.less]: ok

File: contribs/gmf/src/directives/timeslider.js [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+  /**
+   * @type {Object}
+   * @export
+   */
+  this.time;
+
+
+  var min = new Date(this.time.minValue).getTime();
+  var max = new Date(this.time.maxValue).getTime();
+  var step = (max - min) / (1000 * 3600 * 24);
```
Comment (olivierSemet) [inline contribs/gmf/src/directives/timeslider.js]: wrong, should be var step = 1000 * 3600 * 24;

File: contribs/gmf/src/directives/timeslider.js (hunk 6 above)
Comment (olivierSemet) [inline contribs/gmf/src/directives/timeslider.js]: because of https://github.com/angular-ui/ui-slider/blob/master/src/slider.js#L175 
...

File: contribs/gmf/examples/timeslider.html [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    <meta name="viewport"
+          content="initial-scale=1.0, user-scalable=no, width=device-width">
+    <meta name="mobile-web-app-capable" content="yes">
+    <link rel="stylesheet" href="../third-party/jquery-ui/jquery-ui.min.css">
+    <link rel="stylesheet" href="../../../node_modules/bootstrap/dist/css/bootstrap.css">
+    <style>
+      ul {
+        width: 50%;
+      }
+
```
Comment (ger-benjamin) [inline contribs/gmf/examples/timeslider.html]: Useless carriage return

File: contribs/gmf/examples/timeslider.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * @constructor
+ * @param {!angular.Scope} $scope Angular scope.
+ * @ngInject
+ */
+app.MainController = function($scope) {
+
+  /**
+   * @type {Object}
```
Comment (ger-benjamin) [inline contribs/gmf/examples/timeslider.js]: Can you be more precise for this.
Perhaps you'll need to create externs...

**Comment by ger-benjamin** [comment]
Can you be more precise on all yours "@type" (and "@function").
Perhaps you should creates some externs to define types.
Doc like this are useless: http://oliviersemet.github.io/ngeo/time-slider-component/apidoc/gmf.DatePickerController.html

File: contribs/gmf/src/directives/timeslider.js [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    'min' : min,
+    'max' : max,
+    'step' : step
+  };
+
+  /**
+   * @type {Array|number}
+   * @export
+   */
+  this.dates = this.time.mode === 'range' && [min, max] || Math.ceil(min + (max - min) / 2);
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/timeslider.js]: That's hard to read.

Can you write it like this:

````
this.dates = this.time.mode === 'range' ? [min, max] : Math.ceil(min + (max - min) / 2);
````
Or even with if/else...

Let the compilator transforms it into something quick to read for machines.

File: contribs/gmf/src/directives/partials/timeslider.html [hunk 15]
```diff
@@ -0,0 +1,6 @@
+<div class="gmf-time-slider">
+  <div ui-slider="sliderCtrl.sliderOptions" ng-model="sliderCtrl.dates">
+    <span class="ui-slider-handle ui-state-default ui-corner-all" tabindex="0" style="left: 0%;"
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/partials/timeslider.html]: I dont' like to have `style` here. What is the comportment of the slider when the given default value isn't the first slider available value ?

File: contribs/gmf/src/directives/timeslider.js [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+     */
+    function(element, attrs) {
+      var templateUrl = attrs['gmfTimeSliderTemplateUrl'];
+      return templateUrl !== undefined ? templateUrl :
+          gmf.baseTemplateUrl + '/timeslider.html';
+    });
+
+
+/**
+ * Provide a directive to select a single date or a range of dates with a slider
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/timeslider.js]: Others directives give an example of html implementation. Can you provide one ?

File: contribs/gmf/src/directives/timeslider.js [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+      }
+    }
+  };
+};
+
+gmf.module.directive('gmfTimeSlider', gmf.TimeSlider);
+
+
+/**
+ * TimeSliderController - directive conttroller
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/timeslider.js]: typo

File: contribs/gmf/src/directives/timeslider.js [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+ * @param {!angular.Scope} $scope Angular scope.
+ * @constructor
+ * @export
+ * @ngInject
+ * @ngdoc controller
+ * @ngname gmfTimeSliderController
+ */
+gmf.TimeSliderController = function($scope) {
+
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/timeslider.js]: Not regular carriage return between each "code block" (sometime 1cr, sometime 2cr. Please align with others controllers).

**Comment by ger-benjamin** [comment]
Oh and it's wanted that desktop layertree don't have timeslider in this PR ?
https://oliviersemet.github.io/ngeo/time-slider-component/examples/contribs/gmf/apps/desktop/index.html

**Comment by olivierSemet** [comment]
@ger-benjamin soon:

> Demo app: Currently the node with a time parameter set to 'slider' is not reachable but as soon as the demo server is updated with @sbrunner modifications, it should be available.

**Comment by ger-benjamin** [comment]
Sorry, plus, I've read that...

File: contribs/gmf/src/directives/timeslider.js (hunk 14 above)
Comment (olivierSemet) [inline contribs/gmf/src/directives/timeslider.js]: ok

File: contribs/gmf/src/directives/partials/timeslider.html (hunk 15 above)
Comment (olivierSemet) [inline contribs/gmf/src/directives/partials/timeslider.html]: This is useless, I forgot to delete it. Thanks

File: contribs/gmf/src/directives/timeslider.js (hunk 16 above)
Comment (olivierSemet) [inline contribs/gmf/src/directives/timeslider.js]: No, I think there is no need for this directive to change the HTML implementation. I'll remove it

**Comment by olivierSemet** [comment]
@ger-benjamin could you review?

File: contribs/gmf/src/directives/timeslider.js [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+  /**
+   * Options for the ui-slider directive
+   * @type {{range: boolean,
+   * min: number,
+   * max: number,
+   * step: number}}
+   * @export
+   */
+  this.sliderOptions = {
+    'range' : this.time.mode === 'range',
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/timeslider.js]: I think that you can remove can write object keys without quote now.

File: contribs/gmf/examples/timeslider.js [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    maxValue: '2013-12-31T00:00:00Z',
+    minValue: '2006-01-01T00:00:00Z',
+    maxDefValue: null,
+    minDefValue: null,
+    resolution: 'day',
+    mode: 'range'
+  };
+
+  /**
+   * @type {Object}
```
Comment (ger-benjamin) [inline contribs/gmf/examples/timeslider.js]: `{gmfx.TimeProperty}` ?

File: contribs/gmf/examples/timeslider.js [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+    maxValue: '2015-12-31T00:00:00Z',
+    minValue: '2014-01-01T00:00:00Z',
+    maxDefValue: null,
+    minDefValue: null,
+    resolution: 'day',
+    mode: 'single'
+  };
+
+  /**
+   * @type {Object}
```
Comment (ger-benjamin) [inline contribs/gmf/examples/timeslider.js]: `{gmfx.TimeProperty}` ?

File: contribs/gmf/examples/timeslider.js [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+   * @export
+   */
+  this.sliderValue = {};
+
+  /**
+   * @type {Object}
```
Comment (ger-benjamin) [inline contribs/gmf/examples/timeslider.js]: `{gmfx.TimeProperty}` ?

File: contribs/gmf/examples/timeslider.js (hunk 13 above)
Comment (ger-benjamin) [inline contribs/gmf/examples/timeslider.js]: `{gmfx.TimeProperty}` ?

**Comment by ger-benjamin** [comment]
I've don't passed a lot of time on the "link" function in your directive. I hope that's okay but it seems that @pgiraud has already saw it.

Adjust the `type` in example then it's okay for me.

**Comment by olivierSemet** [comment]
@ger-benjamin  thanks for review

File: contribs/gmf/examples/timeslider.js (hunk 21 above)
Comment (olivierSemet) [inline contribs/gmf/examples/timeslider.js]: Not here

File: contribs/gmf/examples/timeslider.js (hunk 22 above)
Comment (olivierSemet) [inline contribs/gmf/examples/timeslider.js]: No

**Comment by ger-benjamin** [comment]
@olivierSemet Thank you, It looks good now. Can I merge ?

**Comment by pgiraud** [comment]
@ger-benjamin Please don't.*
We still have to address the tooltip issues.

**Comment by pgiraud** [comment]
Démo for the desktop application:
https://oliviersemet.github.io/ngeo/time-slider-component/examples/contribs/gmf/apps/desktop

**Comment by olivierSemet** [comment]
WIP

File: contribs/gmf/apps/desktop_alt/js/controller.js [hunk 23]
```diff
@@ -18,6 +18,10 @@ goog.require('gmf.authenticationDirective');
 /** @suppress {extraRequire} */
 goog.require('ngeo.proj.EPSG2056');
 /** @suppress {extraRequire} */
+goog.require('gmf.DatePickerDirective');
+/** @suppress {extraRequire} */
+goog.require('gmf.TimeSliderDirective');
+/** @suppress {extraRequire} */
```
Comment (sbrunner) [inline contribs/gmf/apps/desktop_alt/js/controller.js]: This should be in the abstract desktop controller :-)

**Comment by fredj** [comment]
It would be nice if the time slider looks the same as the other sliders (see `contribs/gmf/less/input-range.less`), it can be done later in an other PR.

File: contribs/gmf/src/directives/timeslider.js [hunk 24]
```diff
@@ -0,0 +1,177 @@
...
+   */
+  this.time;
+
+  var min = new Date(this.time.minValue).getTime();
+  var max = new Date(this.time.maxValue).getTime();
+  var step = 1000 * 3600 * 24; //day resolution
+
+  /**
+   * Options for the ui-slider directive
+   * @type {{range: boolean,
```
Comment (fredj) [inline contribs/gmf/src/directives/timeslider.js]: A `@typedef` should be created in this file (see https://github.com/camptocamp/ngeo/blob/master/src/services/query.js for an example)

**Comment by ybolognini** [comment]
I agree with @fredj, would be nice to have same look for all sliders

**Comment by olivierSemet** [comment]
@fredj @ybolognini  can we close this PR, this is no longer the good one. Check #1422
